### PR TITLE
Product custom options (#121)

### PR DESCRIPTION
**Pull request by rus-daniel**

### CODE REVIEW

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 1]
```diff
@@ -507,6 +517,21 @@ protected function _addAmount(Mage_Catalog_Model_Product $product, OystProduct &
      */
     protected function _getPrices(Mage_Catalog_Model_Product $product, $storeId = null)
     {
+        $customPrice = 0;
+        if ($this->customOptions) {
+            foreach (json_decode($this->customOptions) as $option) {
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: Use ```Zend_Json::decode```

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 2]
```diff
@@ -130,6 +130,11 @@ class Oyst_OneClick_Model_Catalog extends Mage_Core_Model_Abstract
      */
     private $stockItem = null;
 
+    /**
+     * @var null
+     */
+    private $customOptions = null;
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: change to ```$configurableProductCustomOptions```

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 3]
```diff
@@ -212,11 +217,16 @@ public function processNotification($event, $apiData)
      *
      * @param int $productId
      * @param int $configurableProductChildId
+     * @param string $customOptions
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: change to ```$configurableProductCustomOptions```

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 4]
```diff
@@ -212,11 +217,16 @@ public function processNotification($event, $apiData)
...
      *
      * @return OystProduct
      */
-    public function getOystProduct($productId, $configurableProductChildId = null)
+    public function getOystProduct($productId, $configurableProductChildId = null, $customOptions = null)
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: change to ```$configurableProductCustomOptions```

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 5]
```diff
@@ -212,11 +217,16 @@ public function processNotification($event, $apiData)
...
     {
+        if (strlen($customOptions) > 2) {
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: change to ```$configurableProductCustomOptions```

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 6]
```diff
@@ -507,6 +517,21 @@ protected function _addAmount(Mage_Catalog_Model_Product $product, OystProduct &
      */
     protected function _getPrices(Mage_Catalog_Model_Product $product, $storeId = null)
     {
+        $customPrice = 0;
+        if ($this->customOptions) {
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: change to ```$configurableProductCustomOptions```

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 7]
```diff
@@ -507,6 +517,21 @@ protected function _addAmount(Mage_Catalog_Model_Product $product, OystProduct &
...
+            foreach (json_decode($this->customOptions) as $option) {
+                preg_match('!\d+!', $option->optionId, $matches);
+                $customProduct = Mage::getModel('catalog/product')->load($product->getId());
+
+                /** @var Mage_Catalog_Model_Product_Option $customOption */
+                $customOption = $customProduct->getOptionById($matches[0]);
+
+                if ($optionValue = $option->value) {
+                    $customOption = $customOption->getValueById($optionValue);
+                }
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: Break line after ```if```.

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 8]
```diff
@@ -45,6 +45,7 @@ function oystOneClick(productId, childId, oneClickUrl, isProductAddtocartFormVal
 
             form.append("configurableProductChildId", configurableProductChildId);
             form.append("preload", opts.preload);
+            form.append("customOptions", JSON.stringify(getCustomOptions()));
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/oneclick.js]: Change to ```configurableProductCustomOptions``` and change to ```getConfigurableProductCustomOptions()``` for JS.

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 9]
```diff
@@ -163,3 +164,25 @@ function isOystOneClickButtonFormValid(formName) {
 function prependChild(parentElement, newFirstChildElement) {
     parentElement.insertBefore(newFirstChildElement, parentElement.firstChild);
 }
+
+function getCustomOptions() {
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/oneclick.js]: Change to ```getConfigurableProductCustomOptions()```

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 10]
```diff
@@ -507,6 +517,21 @@ protected function _addAmount(Mage_Catalog_Model_Product $product, OystProduct &
      */
     protected function _getPrices(Mage_Catalog_Model_Product $product, $storeId = null)
     {
+        $customPrice = 0;
+        if ($this->customOptions) {
+            foreach (json_decode($this->customOptions) as $option) {
+                preg_match('!\d+!', $option->optionId, $matches);
+                $customProduct = Mage::getModel('catalog/product')->load($product->getId());
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: Add ```// @codingStandardsIgnoreLine``` above this this line to make build ok.

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 11]
```diff
@@ -163,3 +164,24 @@ function isOystOneClickButtonFormValid(formName) {
...
+                value: null
+            };
+            switch (element.nodeName) {
+                case "SELECT":
+                    input.value = element.value;
+                    break;
+            }
+            ids.push(input);
+        }
+    });
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/oneclick.js]: add break line.

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 12]
```diff
@@ -163,3 +164,24 @@ function isOystOneClickButtonFormValid(formName) {
...
 }
+
+function configurableProductCustomOptions() {
+    var customOptions = $$('.product-custom-option'),
+        ids = [];
+
+    customOptions.forEach(function (element) {
+        if (element.value) {
+            var input = {
+                optionId: element.getAttribute('name'),
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/oneclick.js]: In Js use double quote.

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 13]
```diff
@@ -163,3 +164,24 @@ function isOystOneClickButtonFormValid(formName) {
 function prependChild(parentElement, newFirstChildElement) {
     parentElement.insertBefore(newFirstChildElement, parentElement.firstChild);
 }
+
+function configurableProductCustomOptions() {
+    var customOptions = $$('.product-custom-option'),
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/oneclick.js]: Use VanillaJs (document.getElementsby... ) to me more compatible with customer front and in use double quote ".

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 14]
```diff
@@ -212,11 +217,16 @@ public function processNotification($event, $apiData)
...
      * @param int $productId
      * @param int $configurableProductChildId
+     * @param string $configurableProductCustomOptions
      *
      * @return OystProduct
      */
-    public function getOystProduct($productId, $configurableProductChildId = null)
+    public function getOystProduct($productId, $configurableProductChildId = null, $configurableProductCustomOptions = null)
     {
+        if (strlen($configurableProductCustomOptions) > 2) {
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: Use defensive programming in your condition.

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 15]
```diff
@@ -507,6 +517,23 @@ protected function _addAmount(Mage_Catalog_Model_Product $product, OystProduct &
      */
     protected function _getPrices(Mage_Catalog_Model_Product $product, $storeId = null)
     {
+        $customPrice = 0;
+        if ($this->configurableProductCustomOptions) {
+            foreach (Zend_Json::decode($this->configurableProductCustomOptions) as $option) {
+                preg_match('!\d+!', $option->optionId, $matches);
+                $customProduct = Mage::getModel('catalog/product')->load($product->getId());
+
+                // @codingStandardsIgnoreLine
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: ```// @codingStandardsIgnoreLine```
have to be add before 
```$customProduct = Mage::getModel('catalog/product')->load($product->getId());```

image

**Comment by StefBrito** [comment]
There the error that i got :
image

When i put a text in the text field i get this error i can't call the OneClickUrl.

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 16]
```diff
@@ -174,3 +175,44 @@ function oneClickButtonPickToFirstAddToCartButtons() {
         prependChild(addToCartButtons, oystOneClickButton);
     });
 }
+
+function configurableProductCustomOptions() {
+    var customOptions = document.getElementsByClassName("product-custom-option"),
+        inputType, name, optionId, value, options = [];
+
+    for (var i = 0; i < customOptions.length; i++) {
+
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/oneclick.js]: remove this break line

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 17]
```diff
@@ -174,3 +175,44 @@ function oneClickButtonPickToFirstAddToCartButtons() {
...
+        if (customOptions[i].hasAttribute("name")) {
+            inputType = customOptions[i].getAttribute("type");
+            name      = customOptions[i].getAttribute("name");
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/oneclick.js]: remove useless space before =

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 18]
```diff
@@ -174,3 +175,44 @@ function oneClickButtonPickToFirstAddToCartButtons() {
...
 }
+
+function configurableProductCustomOptions() {
+    var customOptions = document.getElementsByClassName("product-custom-option"),
+        inputType, name, optionId, value, options = [];
+
+    for (var i = 0; i < customOptions.length; i++) {
+
+        if (customOptions[i].hasAttribute("name")) {
+            inputType = customOptions[i].getAttribute("type");
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/oneclick.js]: add var for each new variable

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 19]
```diff
@@ -174,3 +175,44 @@ function oneClickButtonPickToFirstAddToCartButtons() {
...
+            name      = customOptions[i].getAttribute("name");
+            value     = customOptions[i].value;
+            optionId  = name.replace(/\D/g,'');
+            var customOption = {
+                optionId: optionId,
+                value: value
+            };
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/oneclick.js]: add break line after

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 20]
```diff
@@ -174,3 +175,44 @@ function oneClickButtonPickToFirstAddToCartButtons() {
         prependChild(addToCartButtons, oystOneClickButton);
     });
 }
+
+function configurableProductCustomOptions() {
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/oneclick.js]: Add documentation above the function

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 21]
```diff
@@ -174,3 +175,44 @@ function oneClickButtonPickToFirstAddToCartButtons() {
...
+                    options.push(customOption);
+                }
+            }
+        }
+    }
+
+    return options;
+}
+
+function hasClass(element, cls) {
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/oneclick.js]: Add documentation above the function

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 22]
```diff
@@ -507,6 +518,23 @@ protected function _addAmount(Mage_Catalog_Model_Product $product, OystProduct &
...
+                /** @var Mage_Catalog_Model_Product_Option $customOption */
+                $customOption = $customProduct->getOptionById($option['optionId']);
+
+                if (is_numeric($option['value'])) {
+                    $customOption = $customOption->getValueById($option['value']);
+                }
+
+                $customPrice += $customOption->getPrice(true);
+            }
+        }
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: add break line.

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 23]
```diff
@@ -507,6 +518,23 @@ protected function _addAmount(Mage_Catalog_Model_Product $product, OystProduct &
      */
     protected function _getPrices(Mage_Catalog_Model_Product $product, $storeId = null)
     {
+        $customPrice = 0;
+
+        if ($this->configurableProductCustomOptions) {
+            // @codingStandardsIgnoreLine
+            $customProduct = Mage::getModel('catalog/product')->load($product->getId());
+            foreach ($this->configurableProductCustomOptions as $option) {
+
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: remove breakline

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 24]
```diff
@@ -212,11 +217,17 @@ public function processNotification($event, $apiData)
...
      * @param int $productId
      * @param int $configurableProductChildId
+     * @param string $configurableProductCustomOptions
      *
      * @return OystProduct
      */
-    public function getOystProduct($productId, $configurableProductChildId = null)
+    public function getOystProduct($productId, $configurableProductChildId = null, $configurableProductCustomOptions = null)
     {
+        $data = Zend_Json::decode($configurableProductCustomOptions);
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: do it like that

```
if ($data = Zend_Json::decode($configurableProductCustomOptions)) {
...
}
```

File: app/code/community/Oyst/OneClick/Model/OneClick/ApiWrapper.php [hunk 25]
```diff
@@ -100,7 +100,8 @@ public function authorizeOrder($dataFormated)
 
         /** @var Oyst_OneClick_Model_Catalog $oystCatalog */
         $oystCatalog = Mage::getModel('oyst_oneclick/catalog');
-        $oystProduct = $oystCatalog->getOystProduct($dataFormated['productId'], $dataFormated['configurableProductChildId']);
+        $oystProduct = $oystCatalog->getOystProduct($dataFormated['productId'], $dataFormated['configurableProductChildId'],
+            isset($dataFormated['customOptions']) ? $dataFormated['customOptions'] : null);
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/OneClick/ApiWrapper.php]: Don't put terner in function params, output this before ```$oystProduct```

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 26]
```diff
@@ -174,3 +175,60 @@ function oneClickButtonPickToFirstAddToCartButtons() {
...
+ *
+ * @returns {Array}
+ */
+function configurableProductCustomOptions() {
+    var customOptions = document.getElementsByClassName("product-custom-option");
+    var inputType;
+    var name;
+    var optionId;
+    var value;
+    var options = [];
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/oneclick.js]: Do it like : 
```var inputType, name, optionId, value, options = [];```

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 27]
```diff
@@ -174,3 +175,60 @@ function oneClickButtonPickToFirstAddToCartButtons() {
...
+            inputType = customOptions[i].getAttribute("type");
+            name = customOptions[i].getAttribute("name");
+            value = customOptions[i].value;
+            optionId = name.replace(/\D/g,'');
+            var customOption = {
+                optionId: optionId,
+                value: value
+            };
+
+            if ((inputType === "checkbox" || inputType === "radio") && customOptions[i].checked && value) {
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/oneclick.js]: invert the comparaison (defensive programming)
```"checkbox" === inputType```

do same with other if in this function

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 28]
```diff
@@ -174,3 +175,60 @@ function oneClickButtonPickToFirstAddToCartButtons() {
...
+    }
+
+    return options;
+}
+
+/**
+ * Check if html element has specified class.
+ *
+ * @param element
+ * @param cls
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/oneclick.js]: new breakline

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 29]
```diff
@@ -507,6 +517,23 @@ protected function _addAmount(Mage_Catalog_Model_Product $product, OystProduct &
      */
     protected function _getPrices(Mage_Catalog_Model_Product $product, $storeId = null)
     {
+        $customPrice = 0;
+
+        if ($this->configurableProductCustomOptions) {
+            // @codingStandardsIgnoreLine
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: is this  ```// @codingStandardsIgnoreLine``` is necessary in this case?

**Pull request closed without merging**